polars/polars-time/src/windows/groupby.rs

```diff
@@ -398,7 +404,21 @@ pub(crate) fn groupby_values_iter_full_lookahead(
 
             let b = Bounds::new(lower, upper);
 
-            debug_assert!(i < time.len());
+            // find starting point of window
+            match inner_loop_needed {
```

Comment:
This still adds a branch. 

Let's do the branch completely before `iter().enumerate()`. That makes all paths optimal.

Reply:
so something like
```rust
    match inner_loop_needed {
        true => {
            time[start_offset..upper_bound]
                .iter()
                .enumerate()
                .map(move |(mut i, lower)| {
                    i += start_offset;
                    let lower = add(&offset, *lower, tz.as_ref())?;
                    let upper = add(&period, lower, tz.as_ref())?;

                    let b = Bounds::new(lower, upper);

                    // find starting point of window
                    for &t in &time[i..] {
                        if b.is_member(t, closed_window) {
                            break;
                        }
                        i += 1;
                    }
                    if i >= time.len() {
                        return Ok((i as IdxSize, 0));
                    }
                    let slice = unsafe { time.get_unchecked(i..) };
                    let len = slice.partition_point(|v| b.is_member(*v, closed_window));

                    Ok((i as IdxSize, len as IdxSize))
                })
        }
        false => time[start_offset..upper_bound]
            .iter()
            .enumerate()
            .map(move |(mut i, lower)| {
                i += start_offset;
                let lower = add(&offset, *lower, tz.as_ref())?;
                let upper = add(&period, lower, tz.as_ref())?;

                let b = Bounds::new(lower, upper);

                debug_assert!(i < time.len());
                let slice = unsafe { time.get_unchecked(i..) };
                let len = slice.partition_point(|v| b.is_member(*v, closed_window));

                Ok((i as IdxSize, len as IdxSize))
            }),
    }
}
```
?

This would give
```rust
    = note: expected struct `std::iter::Map<std::iter::Enumerate<std::slice::Iter<'_, _>>, [closure@/home/marcogorelli/polars-dev/polars/polars-time/src/windows/groupby.rs:401:22: 401:43]>`
               found struct `std::iter::Map<std::iter::Enumerate<std::slice::Iter<'_, _>>, [closure@/home/marcogorelli/polars-dev/polars/polars-time/src/windows/groupby.rs:427:18: 427:39]>`
    = note: no two closures, even if identical, have the same type
    = help: consider boxing your closure and/or using it as a trait object
help: you could change the return type to be a boxed trait object
    |
381 | ) -> Box<dyn Iterator<Item = PolarsResult<(IdxSize, IdxSize)>> + TrustedLen + '_> {
    |      ~~~~~~~                                                                    +
help: if you change the return type to expect trait objects, box the returned expressions
    |
398 ~             Box::new(time[start_offset..upper_bound]
399 |                 .iter()
  ...
421 |                     Ok((i as IdxSize, len as IdxSize))
422 ~                 }))
423 |         }
424 ~         false => Box::new(time[start_offset..upper_bound]
425 |             .iter()
  ...
438 |                 Ok((i as IdxSize, len as IdxSize))
439 ~             })),
    |
```
, is that OK? Wouldn't the perf implications of using `Box` be worse than just matching a boolean?

Reply:
Right, then we can do the branch on the caller location. Meaning we split into 2 functions. We already have done this earlier. We have a `full_lookahead`, a `partial_lookbehind` and a `full_lookbehind`. So now we need 1 extra, I believe?